File: rust_feedforward/src/layer.rs
#![allow(unused)]
use ndarray::{Array1, Array2, Axis};
use ndarray_rand::RandomExt;
// use rand::prelude::*;
use rand_distr::StandardNormal;

use crate::activations;

#[derive(Clone, Debug)]
pub enum Activation {
    ReLU,
    Softmax,
}

pub struct Layer {
    pub weights: Array2<f32>,
    pub biases: Array1<f32>,
    pub activation: Activation,

    pub last_input: Option<Array2<f32>>,
    pub last_z: Option<Array2<f32>>,
}

impl Layer {
    pub fn new(input_dim: usize, output_dim: usize, activation: Activation) -> Self {
        // using a normal distribution.
        let weights = Array2::<f32>::random((input_dim, output_dim), StandardNormal);
        let biases = Array1::<f32>::zeros(output_dim);
        Self {
            weights,
            biases,
            activation,
            last_input: None,
            last_z: None,
        }
    }

    pub fn forward(&mut self, input: &Array2<f32>) -> Array2<f32> {
        self.last_input = Some(input.clone());
        let z = input.dot(&self.weights) + &self.biases;
        self.last_z = Some(z.clone());
        crate::activations::apply(&self.activation, &z)
    }

    pub fn backward(&mut self, d_a: &Array2<f32>, lr: f32) -> Array2<f32> {
        let z = self.last_z.as_ref().expect("Cached z not found");
        let input = self.last_input.as_ref().expect("Cached input not found");

        // derivative of the activation function.
        let d_activation = activations::derivative(&self.activation, z);

        // combine with the gradient coming from the next layer.
        let d_z = d_a * &d_activation;

        // gradients with respect to weights and biases.
        let d_w = input.t().dot(&d_z);
        let db = d_z.sum_axis(Axis(0));

        // the gradient to the previous layer.
        let d_x = d_z.dot(&self.weights.t());

        self.weights -= &(lr * &d_w);
        self.biases -= &(lr * &db);

        d_x
    }
}

File: python_feedforward/src/layer.py
import numpy as np
from activations import *

class Layer:
    def __init__(self, input_dim, output_dim, activation='relu'):
        self.weights = np.random.randn(input_dim, output_dim).astype(np.float32)
        self.biases = np.zeros((1, output_dim), dtype=np.float32)
        self.activation = activation.lower()
        self.last_input = None
        self.last_z = None

    def forward(self, input):
        self.last_input = input
        self.last_z = np.dot(input, self.weights) + self.biases
        
        if self.activation == 'relu':
            return relu(self.last_z)
        elif self.activation == 'softmax':
            return softmax(self.last_z)
        else:
            raise ValueError(f"Unknown activation function: {self.activation}")

    def backward(self, d_a, learning_rate):
        if self.activation == 'relu':
            d_activation = relu_derivative(self.last_z)
        elif self.activation == 'softmax':
            d_activation = softmax_derivative(self.last_z)
        else:
            raise ValueError(f"Unknown activation function: {self.activation}")
        
        d_z = d_a * d_activation
        
        d_w = np.dot(self.last_input.T, d_z)
        d_b = np.sum(d_z, axis=0, keepdims=True)
        
        d_x = np.dot(d_z, self.weights.T)
        
        self.weights -= learning_rate * d_w
        self.biases -= learning_rate * d_b
        
        return d_x
